### PR TITLE
[framework] improve Notification API performance by skipping full Element tree traversal

### LINKED ISSUE #97849: Notifcation API is expensive
From some early investigation, several of the scroll notification events have no listeners if there is no scroll bar, which leads to 500/1000 element traversals and 2x-3x as many typechecks per frame on gallery Shrine/ mobile.

Will fill in more details later

### PR DESCRIPTION
**Pull request by jonahwilliams**
### Background

The notification API must traverse the element tree in order to locate a notification receiver. This can lead to some unfortunate performance characteristics:

https://github.com/flutter/flutter/blob/9407700aa4a843885d7acca324773f85e3766743/packages/flutter/lib/src/widgets/notification_listener.dart#L53-L73

 * If there is no receiver for a given notification type, the entire element tree above the notification will be traversed and type checked.
 * For multiple notifications in a given frame (which is common for scroll views) we can end up traversing the element tree multiple times
 * If there are multiple or nested scroll views on a given page, the situation is worsened significantly.
 
 What is the impact of these performance problems? If we look at devtools as an example app we find that about 30% CPU time while scrolling the flame graph is spent dispatching notifications. Less severely, in flutter/gallery Crane demo on mobile we spend 2-3% CPU dispatching notifications for scrollbars that are entirely unused.
 
 
image

 
 ### Detailed Design
 
 In order to reduce the cost of dispatching notifications, we must traverse fewer elements. This can be done by creating a new Tree (Notification Tree) hosted in the element tree that only contains Elements which can respond to notifications. Since there are many times fewer notification listeners than elements, this is significantly faster.
 
 However, the old notification system exposed the fact that it traversed each element as part of its API. This was used by certain scroll notifications to check for Viewport render objects in order to increment a depth property. If we skip most elements, then we would also skip the corresponding RenderObjectElements. 
 
 https://github.com/flutter/flutter/blob/9407700aa4a843885d7acca324773f85e3766743/packages/flutter/lib/src/widgets/scroll_notification.dart#L27-L33
 
 To support this, we need to not only create an Element type for NotificationListeners, but provide an API that allows different Element subtypes to subscribe to notifications. The following is the new Element API added:
 
 ```dart
/// Mixin this class to Elements allow receiving [Notification] objects dispatched
/// by children.
///
/// See also:
///   * [NotificationListener], for a widget that allows consuming notifications.
mixin NotifiableElementMixin on Element {
  /// Return true to consume the notification or false to let it continue bubbling.
  bool onNotification(Notification notification);

}

```
 
 ### Improvements
 
 The result of this change is that notification dispatch drops to about 2-3% CPU on devtools and no longer appears in Crane. This is likely to be a breaking change, though I have not yet run google3 tests to confirm. Regardless, we may need to consider a migrating strategy where for a period of time both dispatching mechanisms are supported
 
 
image

Fixes https://github.com/flutter/flutter/issues/97849

### CODE REVIEW

**Comment by goderbauer** [comment]
It would be cool if ViewportBoundary could just be an implementation detail of the Viewport widget (without the need to expose something like ViewportInternal)... As an idea: could the `Viewport` widget's custom element `_ViewportElement` just override `dispatchNotification` with the depth increasing logic?

**Comment by jonahwilliams** [comment]
dispatchNotification would skip any Element that isn't a notification element by design. We'd have to expose some APIs for overriding that allow other element types to participate in the notification tree.

I think you're right though, this is probably the way to go and involves fewer breaking changes.

I'll think about this one some more....

**Comment by jonahwilliams** [comment]
Tried to do some stuff with mixins but it is quite tricky.

**Comment by jonahwilliams** [comment]
OK!

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -4361,6 +4404,21 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
...
+  @override
+  bool onNotification(Notification notification) {
+    return false;
+  }
+
+  /// Whether this element can handle [Notifications].
+  ///
+  /// The value of this getter must be unchanging for the lifetime of the Element.
+  bool get handlesNotification => false;
+
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/widgets/framework.dart:4421]: @goderbauer this is basically the new API I added to make this work. I'm not 100% how I feel about it, but it does allow the viewport elements to just mixin a class to support the scroll notifications

**Comment by jonahwilliams** [comment]
Need to fill out docs more obvs...

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart:4421]: You mentioned you tried mixin's, but could we have something like this:

* `Element._attachNotificationTree` always sets `_notificationTree = _parent?._notificationTree;`
* a new `NotifiableElementMixin` overrides `_attachNotificationTree` to set itself as the `_notificationTree` and it provides the abstract `onNotification` method.
* `Element.onNotification` is removed

That way we wouldn't need the awkward `handlesNotification` thing...

Users would just create custom elements with `NotifiableElementMixin` mixed in and implement the onNotification method.

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:4421]: You can't apply a mixin to a type that overrides a private method. For example, the _attachNotificationTree override defined by the mixin will conflict with the _attachNotificationTree that the NotificationElement inherits from Element.

The only way around this is to make the method public. We could use a protected annotation though. I do like this API better than what I have.

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart:4421]: > You can't apply a mixin to a type that overrides a private method.

This seems to work in dartpad:

```dart
class Foo {
  foo() {
    _attachNotificationTree();
  }
  
  _attachNotificationTree() {
    print("hello");
  }
}

mixin FooMixin on Foo {
  @override
  _attachNotificationTree() {
    print("Hell from mixin");
  }
}

class Ha extends Foo with FooMixin {
  
}

main() {
  Ha().foo();
}
```

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart]: Ugh, never mind. This only works because Ha is in the same file...

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart]: Making it protected doesn't sound like the worst idea then...

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart]: Yeah, its not so bad - and its not the only element API that is protected too

File: packages/flutter/lib/src/widgets/scroll_notification.dart [hunk 2]
```diff
@@ -24,20 +24,29 @@ mixin ViewportNotificationMixin on Notification {
...
   }
 }
 
+/// A mixin that allows [Element]s containing Viewport like widgets to correctly
+/// modify the notification depth of a [ViewportNotificationMixin].
+///
+/// See also:
+///   * [Viewport], which creates a custom [MultiChildRenderObjectElement] that mixes
+///     this in.
+mixin ViewportElement on NotifiableElementMixin {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_notification.dart:40]: We should probably link to this from some other docs as well so people implementing custom viewports find it.

**Review by goderbauer — COMMENTED** [review]
now the big question: Does this break anything in e.g. google3?

**Comment by jonahwilliams** [comment]
I'm running a full presubmit right now 👯

File: packages/flutter/lib/src/widgets/framework.dart [hunk 3]
```diff
@@ -3103,6 +3111,35 @@ class BuildOwner {
   }
 }
 
+/// Mixin this class to Elements allow receiving [Notification] objects dispatched
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart]: fixed

File: packages/flutter/lib/src/widgets/framework.dart [hunk 4]
```diff
@@ -3103,6 +3111,35 @@ class BuildOwner {
...
+/// by children.
+///
+/// See also:
+///   * [NotificationListener], for a widget that allows consuming notifications.
+mixin NotifiableElementMixin on Element {
+  /// Return true to consume the notification or false to let it continue bubbling.
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart]: borrowed some existing docs

File: packages/flutter/lib/src/widgets/framework.dart [hunk 5]
```diff
@@ -3164,6 +3201,9 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
   Element? _parent;
   DebugReassembleConfig? _debugReassembleConfig;
 
+  _NotificationNode? _notificationTree;
+
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:3208]: done

File: packages/flutter/lib/src/widgets/notification_listener.dart [hunk 6]
```diff
@@ -146,16 +116,30 @@ class NotificationListener<T extends Notification> extends StatelessWidget {
...
-      final bool result = onNotification!(notification);
-      return result == true; // so that null and false have the same effect
+  @override
+  Element createElement() {
+    return _NotificationElement<T>(this);
+  }
+}
+
+/// An element used to host [NotificationListener] elements.
+@optionalTypeArgs
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/notification_listener.dart]: outdated, removed

File: packages/flutter/lib/src/widgets/scroll_notification.dart [hunk 7]
```diff
@@ -24,20 +24,29 @@ mixin ViewportNotificationMixin on Notification {
...
-  }
-
   @override
   void debugFillDescription(List<String> description) {
     super.debugFillDescription(description);
     description.add('depth: $depth (${ depth == 0 ? "local" : "remote"})');
   }
 }
 
+/// A mixin that allows [Element]s containing Viewport like widgets to correctly
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/scroll_notification.dart]: done

File: packages/flutter/lib/src/widgets/scroll_notification.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/scroll_notification.dart]: done

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/scroll_notification.dart]: What is a good place where we describe this - on Viewport itself?

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/scroll_notification.dart]: added a See More section

File: packages/flutter/lib/src/widgets/notification_listener.dart [hunk 8]
```diff
@@ -50,37 +50,14 @@ abstract class Notification {
...
-  ///
-  /// The [Notification] class implementation of this method dispatches the
-  /// given [Notification] to each ancestor [NotificationListener] widget.
-  ///
-  /// Subclasses can override this to apply additional filtering or to update
-  /// the notification as it is bubbled (for example, increasing a `depth` field
-  /// for each ancestor of a particular type).
-  @protected
-  @mustCallSuper
-  bool visitAncestor(Element element) {
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/widgets/notification_listener.dart:63]: Even though nothing broke in customer tests or g3, do you think we should leave this and mark it deprecated? I could also leave the old dispatch path and instead add a new method to do a fast dispatch

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/notification_listener.dart:63]: leaving this and the old dispatch in as `deprecated` sounds reasonable to me. I just wonder if we can find a reasonable new name for the new dispatch. `fastDispatch` sounds strange after we remove the regular `dispatch` - and what if we find an even faster dispatch in the future?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/notification_listener.dart (hunk 8 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/notification_listener.dart:63]: maybe I could _just_ leave the BuildContext dispatch and deprecate the dispatch on Notification?

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/notification_listener.dart:63]: that is, BuildContext would have the new dispatch and Notification would keep the old

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/notification_listener.dart:63]: That sounds reasonable. We'd just have to check all docs pointing to it on the Notification and update them to point to BuildContext instead.

**Comment by jonahwilliams** [comment]
I did spend some time going through the migration guide today. I guess I'm sort of concerned that for folks that don't override visitAncestor on Notification ... we're going to create a bunch of churn for no reason. Maybe I need to do a package survey to see if anyone actually does

**Comment by jonahwilliams** [comment]
I looked through google3 and while I found several notification classes none of them override the visit ancestor functionality. I am going to change this back to it was to reduce churn and we can re-evaluate if anyone flags it as an issue

File: packages/flutter/lib/src/widgets/notification_listener.dart [hunk 9]
```diff
@@ -60,13 +60,13 @@ abstract class Notification {
   /// for each ancestor of a particular type).
   @protected
   @mustCallSuper
+  @Deprecated(
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/widgets/notification_listener.dart:63]: I'd like to leave this deprecated just in case. Would it be appropriate to add a code snippet for how to replace this, or I could add some sort of "dispatch slowly"

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/notification_listener.dart:84]: PTAL, added a code snippet here. Could also add it to a migration guide

File: packages/flutter/lib/src/widgets/notification_listener.dart [hunk 10]
```diff
@@ -58,15 +58,36 @@ abstract class Notification {
...
+  /// This functionality was deprecated after v1.11.0 because of poor performance.
+  /// If you have a Notification subtype that overrides this method, you can instead
+  /// manually implement the old-style notification dispatch with the following
+  /// snippet:
+  ///
+  /// ```dart
+  ///  void dispatchNotificationSlowly<T extends Notification>(T notification, BuildContext context) {
+  ///    context.visitAncestorElements((Element element) {
+  ///       // Perform visit ancestor logic.
+  ///       notification.myAncestorLogic(element);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/notification_listener.dart:71]: It's not super clear what this is. Maybe say explicitly that this is were the custom filtering/update logic (mentioned in one of the paragraphs above) would go that prior to the deprecation went into `visitAncestor`?

**Comment by goderbauer** [comment]
I am slightly confused by this now, but maybe I misunderstood the original plan. I thought the idea was to deprecate `Notification.dispatch` and `Notification.visitAncestor`, but keeping their implementation untouched. In the deprecation notice, we would redirect people to the new `BuildContext.dispatchNotification` and tell them what to do if they want filtering logic that previously went into `visitAncestor`. 

As it stands, this PR now redirects `Notification.dispatch` directly to `BuildContext.dispatchNotification`, bypassing `Notification.visitAncestor` altogether. So, if I relied on `visitAncestor` I'd still be broken - even worse, it would likely be a runtime breakage.

**Comment by goderbauer** [comment]
Oh, I think I missed https://github.com/flutter/flutter/pull/98451#issuecomment-1048931199. That makes sense then. Let's be very clear in the docs to `visitAncestor` that this method is no longer used. Implementing it has no effect since it is deprecated.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart [hunk 11]
```diff
@@ -526,13 +526,13 @@ class _DraggableSheetExtent {
   void updateSize(double newSize, BuildContext context) {
     assert(newSize != null);
     _currentSize.value = newSize.clamp(minSize, maxSize);
-    DraggableScrollableNotification(
+    context.dispatchNotification(DraggableScrollableNotification(
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:529]: revert this change since calling `dispatch` continues to be valid?

File: packages/flutter/lib/src/widgets/notification_listener.dart [hunk 12]
```diff
@@ -58,15 +58,36 @@ abstract class Notification {
   /// Subclasses can override this to apply additional filtering or to update
   /// the notification as it is bubbled (for example, increasing a `depth` field
   /// for each ancestor of a particular type).
+  ///
+  /// This functionality was deprecated after v1.11.0 because of poor performance.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/notification_listener.dart:62]: Update the docs above to make it very clear that this is not only deprecated, but the feature has already been removed. Overriding this method has no effect whatsover.

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/notification_listener.dart:62]: Actually, I am wondering if we should just remove this altogether then. Seeing this deprecated gives you the false sense that - while this will be removed in the future - it still functions as expected. It just confused myself when I took another look at this PR (see my top-level comment).

The code snippet should then go into a migration guide on our website that people will hopefully find when they get invalid override warnings.

File: packages/flutter/lib/src/widgets/notification_listener.dart [hunk 13]
```diff
@@ -111,31 +130,21 @@ abstract class Notification {
 /// Notifications will trigger the [onNotification] callback only if their
 /// [runtimeType] is a subtype of `T`.
 ///
-/// To dispatch notifications, use the [Notification.dispatch] method.
-class NotificationListener<T extends Notification> extends StatelessWidget {
+/// To dispatch notifications, use the [BuildContext.dispatchNotification] method.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/notification_listener.dart:133]: Revert this since the old path is still valid now?

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart [hunk 14]
```diff
@@ -233,7 +233,7 @@ class _GlowingOverscrollIndicatorState extends State<GlowingOverscrollIndicator>
       final bool isLeading = controller == _leadingController;
       if (_lastNotificationType is! OverscrollNotification) {
         final OverscrollIndicatorNotification confirmationNotification = OverscrollIndicatorNotification(leading: isLeading);
-        confirmationNotification.dispatch(context);
+        context.dispatchNotification(confirmationNotification);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/overscroll_indicator.dart:236]: here and elsewhere: revert this since it's just unnecessary churn now?

**Comment by jonahwilliams** [comment]
Sorry for the confusion - yes - since all of the subclasses of notification I looked at do not override visit Ancestor, I'm worried about additional churn when we don't really need to make the change a compile error.

File: packages/flutter/lib/src/widgets/notification_listener.dart (hunk 12 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/notification_listener.dart:62]: OK, that sounds good. I guess its not a compile error to override stuff that isn't there, just a warning

**Comment by jonahwilliams** [comment]
TODO migration guide:

```
  /// Applied to each ancestor of the [dispatch] target.
  ///
  /// The [Notification] class implementation of this method dispatches the
  /// given [Notification] to each ancestor [NotificationListener] widget.
  ///
  /// Subclasses can override this to apply additional filtering or to update
  /// the notification as it is bubbled (for example, increasing a `depth` field
  /// for each ancestor of a particular type).
  ///
  /// This functionality was deprecated after v1.11.0 because of poor performance.
  /// If you have a Notification subtype that overrides this method, you can instead
  /// manually implement the old-style notification dispatch with the following
  /// snippet:
  ///
  /// ```dart
  ///  void dispatchNotificationSlowly<T extends Notification>(T notification, BuildContext context) {
  ///    context.visitAncestorElements((Element element) {
  ///       // Perform visit ancestor logic.
  ///       notification.myAncestorLogic(element);
  ///       if (element is ProxyElement) {
  ///         final Widget widget = element.widget;
  ///         if (widget is NotificationListener<T>) {
  ///           return widget.onNotification?.call(notification) ?? true;
  ///         }
  ///       }
  ///       return true;
  ///    });
  ///  }
  /// ```
```

File: packages/flutter/lib/src/widgets/notification_listener.dart (hunk 13 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/notification_listener.dart:91]: done

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart (hunk 14 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/overscroll_indicator.dart]: done

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart (hunk 11 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:529]: done

**Comment by jonahwilliams** [comment]
Updated this PR by deleting the visitAncestor method - i'll use the notes I collected for a migration guide

File: packages/flutter/lib/src/widgets/notification_listener.dart (hunk 13 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/notification_listener.dart:91]: > done

This appears to be a lie 😝  (here and `scroll_activity.dart` as well).

Ultimately, of course it doesn't matter. But I also don't see a good argument to switch our recommendation.

**Review by goderbauer — APPROVED** [review]
LGTM once we have a migration guide.

**Comment by jonahwilliams** [comment]
for some reason my changes disappeared, sorry! Let me find them....

**Comment by jonahwilliams** [comment]
Draft migration guide at https://github.com/flutter/website/pull/6865
